File: pre-datathon/tools/error_metric.R


error_metric <- function(real, prediction, upper_bound, lower_bound) {
  
  real_tilde <- pmax(real, 1)
  
  general_term <- (abs(real - prediction) + abs(real - upper_bound) + abs(real - lower_bound)) / real_tilde
  
  relative_term_upper <- (1.5 + (2 * abs(real - upper_bound) / real_tilde))
  relative_term_lower <- (1.5 + (2 * abs(real - lower_bound) / real_tilde))
  
  
  relative_term <- rep(1, length(general_term))
  relative_term[real > upper_bound] <- relative_term_upper[real > upper_bound]
  
  relative_term[real < lower_bound] <- relative_term_lower[real < lower_bound]
  print("Relative term")
  print(mean(relative_term))
  print("General term")
  print(mean(general_term))
  
  mean(relative_term * general_term)
  
}


# real = 0 * 1:100 
# prediction = real + 1
# upper_bound = prediction + 0:99
# lower_bound = prediction - 0:99

# print(error_metric(real, prediction, upper_bound, lower_bound))

File: pre-datathon/tools/error_metric.py

import numpy as np


def error_metric(real, prediction, upper_bound, lower_bound):

  real_tilde = np.maximum(real, 1)

  err_pred = np.abs(real - prediction)
  err_upper = np.abs(real - upper_bound)
  err_lower = np.abs(real - lower_bound)
  general_term =  (err_pred + err_lower + err_upper) / real_tilde

  relative_upper = (1.5 + (2 * err_upper / real_tilde))
  relative_lower = (1.5 + (2 * err_lower / real_tilde))

  relative_error = np.ones(len(real))

  relative_error[real > upper_bound] = relative_upper[real > upper_bound]
  relative_error[real < lower_bound] = relative_lower[real < lower_bound]

  print(f"Relative term {np.mean(relative_error)}")
  print(f"General term {np.mean(general_term)}")

  return np.mean(relative_error * general_term)


# real = np.zeros(100)
# prediction = real + 1
# upper_bound = prediction + np.arange(100)
# lower_bound = prediction - np.arange(100)

# print(error_metric(real, prediction, upper_bound, lower_bound))